### PR TITLE
Fix exception during startup when kar is not yet loaded

### LINKED ISSUE #2571: Exceptions when features are installed before KAR installation completed
Major issue with snapshot 2441 (first startup), apparently a problem to install rrd4j.
What is strange is that I am using the kar file which normally contains everything and so nothing should require a download. This looks like a "corrupted" kar file.
```
org.apache.karaf.features.internal.util.MultiException: Error:
        Error downloading mvn:org.openhab.addons.bundles/org.openhab.persistence.rrd4j/3.1.0-SNAPSHOT
        at org.apache.karaf.features.internal.download.impl.MavenDownloadManager$MavenDownloader.<init>(MavenDownloadManager.java:91)
        at org.apache.karaf.features.internal.download.impl.MavenDownloadManager.createDownloader(MavenDownloadManager.java:72)
        at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:457)
        at org.apache.karaf.features.internal.region.Subsystem.downloadBundles(Subsystem.java:452)
        at org.apache.karaf.features.internal.region.SubsystemResolver.resolve(SubsystemResolver.java:224)
        at org.apache.karaf.features.internal.service.Deployer.deploy(Deployer.java:399)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.doProvision(FeaturesServiceImpl.java:1069)
        at org.apache.karaf.features.internal.service.FeaturesServiceImpl.lambda$doProvisionInThread$13(FeaturesServiceImpl.java:1004)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
        at java.base/java.lang.Thread.run(Thread.java:834)
        Suppressed: java.io.IOException: Error downloading mvn:org.openhab.addons.bundles/org.openhab.persistence.rrd4j/3.1.0-SNAPSHOT
                at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:77)
                at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
                at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
                at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
                ... 3 more
        Caused by: java.io.IOException: Error resolving artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT: [Could not find artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT]
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.configureIOException(AetherBasedResolver.java:803)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:774)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:657)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:598)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:565)
                at org.apache.karaf.features.internal.download.impl.MavenDownloadTask.download(MavenDownloadTask.java:52)
                at org.apache.karaf.features.internal.download.impl.AbstractRetryableDownloadTask.run(AbstractRetryableDownloadTask.java:60)
                ... 6 more
                Suppressed: shaded.org.eclipse.aether.transfer.ArtifactNotFoundException: Could not find artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:403)
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215)
                        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192)
                        at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247)
                        at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767)
                        ... 11 more
        Caused by: shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Error resolving artifact org.openhab.addons.bundles:org.openhab.persistence.rrd4j:jar:3.1.0-SNAPSHOT
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:413)
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:215)
                at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:192)
                at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:247)
                at org.ops4j.pax.url.mvn.internal.AetherBasedResolver.resolve(AetherBasedResolver.java:767)
                ... 11 more
```
@kaikreuzer : certainly something critical for the final 3.1.

### PR DESCRIPTION
**Pull request by J-N-K**
Fixes #2571 

I can't test it, though (I was not able to reproduce the issue). Is there an easy way to reproduce the problem? @lolodomo I can provide instructions for a test, if you can reproduce it.

Signed-off-by: Jan N. Klug <github@klug.nrw>

### CODE REVIEW

**Comment by lolodomo** [comment]
> Is there an easy way to reproduce the problem?

For me, using a RPI3, I can reproduce systematically the problem at the very first start of OH (start from scratch).

**Comment by J-N-K** [comment]
It's a bit difficult without building a full distribution, is cleaning cache and tmp also causing it? Do you have addons configured in addons.cfg or via UI?

File: bundles/org.openhab.core.karaf/pom.xml [hunk 1]
```diff
@@ -49,6 +49,12 @@
       <version>${karaf.compile.version}</version>
       <scope>provided</scope>
     </dependency>
+    <dependency>
+      <groupId>org.apache.karaf.kar</groupId>
+      <artifactId>org.apache.karaf.kar.core</artifactId>
+      <version>${karaf.tooling.version}</version>
+      <scope>compile</scope>
```
Comment (wborn) [inline bundles/org.openhab.core.karaf/pom.xml:56]:
```suggestion
      <version>${karaf.compile.version}</version>
      <scope>provided</scope>
```

**Review by wborn — COMMENTED** [review]
Thanks for looking into this!
I have a small suggestion. :slightly_smiling_face:

File: bundles/org.openhab.core.karaf/pom.xml (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.core.karaf/pom.xml]: Done. I have a question: Why is the `EventPublisher` a static member? If it was an instance member we could omit the null-checks.

Comment (wborn) [reply bundles/org.openhab.core.karaf/pom.xml]: I don't know of a good reason why it should be static either.
Maybe @kaikreuzer remembers why it was added like this in #81.

**Comment by J-N-K** [comment]
@lolodomo I still can't reproduce it. Maybe my test setup is not complex enough or my system is too fast.

Can you setup a test, replace `runtime/system/org/openhab/core/bundles/org.openhab.core.karaf/3.3.0-SNAPSHOT/org.openhab.core.karaf-3.3.0-SNAPSHOT.jar` with this file: https://janessa.me/esh/org.openhab.core.karaf-3.3.0-SNAPSHOT.jar and then start it? If everything works well, you should not see the exceptions but a log message `Some .kar files are not installed yet. Delaying add-on installation by 15s.`.

**Comment by lolodomo** [comment]
Just tried with snapshot 2753 and your updated jar. It seems to work. At first startup, this is very very long but finally I see no errors in console and everything is finally started. I do not see your message "Some .kar files are not installed yet. Delaying add-on installation by 15s.". Look at the first entries in my logs:
```
2022-02-14 23:18:03.630 [INFO ] [org.openhab.ui.internal.UIService   ] - Started UI on port 8080
2022-02-14 23:20:02.662 [WARN ] [org.openhab.core.net.NetUtil        ] - Found multiple local interfaces - ignoring 192.168.1.5
2022-02-14 23:20:04.722 [INFO ] [.core.model.lsp.internal.ModelServer] - Started Language Server Protocol (LSP) service on port 5007
```
First I have to wait to be able to run log:tail in the console, this is not new. Then, I have a first entry and then the next entry is coming only 2 minutes later and then everything starts properly.

No problem at second startup.

Looks like your fixed the problem.

I just don't know if this is normal to wait 2 minutes.

**Comment by wborn** [comment]
>  I do not see your message "Some .kar files are not installed yet. Delaying add-on installation by 15s."

I just tested this PR on a local build and did see the message.

So when the merge conflict is resolved I think we can merge it. :slightly_smiling_face:

**Comment by J-N-K** [comment]
done

**Review by wborn — APPROVED** [review]
Many thanks!

**Pull request merged**